src/Scope.zig

```diff
@@ -219,6 +233,40 @@ pub const Root = struct {
     pub fn contains(root: *Root, name: []const u8) bool {
         return root.containsNow(name) or root.translator.global_names.contains(name) or root.translator.weak_global_names.contains(name);
     }
+
+    pub fn addContainerDecl(root: *Root, container_qt: aro.QualType, decl_node: ast.Node) !void {
+        const tag = decl_node.tag();
+        std.debug.assert(tag == .var_simple or tag == .pub_var_simple);
+        const container_decl = switch (tag) {
+            inline .var_simple, .pub_var_simple => |t| decl_node.castTag(t).?.data.init,
+            else => unreachable,
+        };
+
+        const member_fns = std.ArrayListUnmanaged(*ast.Payload.Func).empty;
+        const container_member_fns = ContainerMemberFns{
+            .container_decl = container_decl,
+            .member_fns = member_fns,
+        };
+        try root.container_member_fns_map.put(root.translator.gpa, container_qt, container_member_fns);
+    }
+
+    pub fn addMemberFunction(root: *Root, func_ty: aro.Type.Func, func: *ast.Payload.Func) !void {
+        if (func.data.name == null) return;
+        if (func_ty.params.len == 0) return;
+        const param1_info = func_ty.params[0];
+        const param1_qt = param1_info.qt;
+
+        const container_qt = loop: switch (param1_qt.type(root.translator.comp)) {
+            .pointer => |pointer| continue :loop pointer.child.type(root.translator.comp),
+            .typedef => |typedef| typedef.base,
+            .typeof => |typeof| typeof.base,
+            else => param1_qt,
+        };
```

Comment:
```suggestion
        std.debug.assert(func.data.name != null);
        if (func_ty.params.len == 0) return;
        
        const param1_base = func_ty.params[0].qt.base(root.translator.comp);
        const container_qt = if (param1_base.type == .pointer)
            param1_base.type.pointer.child.base(root.translator.comp).qt
        else
            param1_base.qt;
```

The first parameter of a member function can only have one level of indirection.